File: C++/17.电话号码的字母组合.cpp
/*
 * @lc app=leetcode.cn id=17 lang=cpp
 *
 * [17] 电话号码的字母组合
 */

// @lc code=start
class Solution
{
public:
    void letterCombinationsDFS(string &digits, vector<string> &dict, int level, string out, vector<string> &res)
    {
        if (level == digits.size())
        {
            res.push_back(out);
            return;
        }
        string str = dict[digits[level] - '0'];
        for (int i = 0; i < str.size(); ++i)
        {
            letterCombinationsDFS(digits, dict, level + 1, out + str[i], res);
        }
    }
    vector<string> letterCombinations(string digits)
    {
        if (digits.empty())
            return {};
        vector<string> res;
        vector<string> dict{"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        letterCombinationsDFS(digits, dict, 0, "", res);
        return res;
    }
};
// @lc code=end

File: JavaScript/17.电话号码的字母组合.js
/*
 * @lc app=leetcode.cn id=17 lang=javascript
 *
 * [17] 电话号码的字母组合
 */

// @lc code=start
/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function (digits) {
  if (digits === "") {
    return [];
  }
  const combinations = [];
  backtrack(digits, 0, "", combinations);
  return combinations;
};

function backtrack(digits, index, currentCombination, combinations) {
  if (index === digits.length) {
    combinations.push(currentCombination);
    return;
  }
  const letters = phoneMap[digits[index]];
  for (const letter of letters) {
    backtrack(digits, index + 1, currentCombination + letter, combinations);
  }
}

const phoneMap = {
  2: ["a", "b", "c"],
  3: ["d", "e", "f"],
  4: ["g", "h", "i"],
  5: ["j", "k", "l"],
  6: ["m", "n", "o"],
  7: ["p", "q", "r", "s"],
  8: ["t", "u", "v"],
  9: ["w", "x", "y", "z"],
};

// @lc code=end
